### PR TITLE
Feat(cnpj): Add CNPJ Type

### PR DESCRIPTION
**Pull request by azlev**
Basic implementation from CNPJ Type.

### CODE REVIEW

File: cnpj.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+				 errmsg("invalid CNPJ"),
+				 errdetail("CNPJ should be less than 99999999999999.")));
+	}
+	check_digits = compute_cnpj_check_digits(cnpj);
+	if (check_digits != (cnpj % 100))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_WRONG_OBJECT_TYPE),
+				 errmsg("invalid CNPJ"),
+				 errdetail("Invalid check digit for the CNPJ. (%02d != %02ld)", check_digits, cnpj % 100)));
```
Comment (fabriziomello) [inline cnpj.c]:
```suggestion
				 errdetail("Invalid check digit for the CNPJ.")));
```
IMHO we don't need to show the right check digits for a given CNPJ.

File: cnpj.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+Datum
+cnpjin(PG_FUNCTION_ARGS)
+{
+	char	   *str = PG_GETARG_CSTRING(0);
+	int			a,
+				b,
+				c,
+				d,
+				e;
```
Comment (fabriziomello) [inline cnpj.c]:
```suggestion
	int			group[5];
```
IMHO using an array the code is more clear and readable.

File: cnpj.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	int64		value = 0;
+
+	if (strlen(str) != 18)
+		value = DirectFunctionCall1(int8in, PG_GETARG_DATUM(0));
+	else
+	{
+		int			count;
+
+		count = sscanf(str, "%d.%d.%d/%d-%d", &a, &b, &c, &d, &e);
```
Comment (fabriziomello) [inline cnpj.c]:
```suggestion
		count = sscanf(str, "%02d.%03d.%03d/%04d-%02d", &group[0], &group[1], &group[2], &group[3], &group[4]);
```
For correctness just get the proper number of digits for each group.

File: cnpj.c [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		if (count != 5)
+			ereport(ERROR,
+					(errcode(ERRCODE_WRONG_OBJECT_TYPE),
+					 errmsg("invalid CNPJ"),
+					 errdetail("CNPJ must be a number or match the XX.XXX.XXX/YYYY-ZZ pattern")));
```
Comment (fabriziomello) [inline cnpj.c]:
```suggestion
					 errdetail("Formatted CNPJ should match the 99.999.999/9999-99 pattern.")));
```

File: expected/cnpj_1.out [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+----------
+ f
+(1 row)
+
+SELECT CNPJ '00000000000000' < CNPJ '00.001.180/0001-26';
+ ?column? 
+----------
+ t
+(1 row)
+
```
Comment (fabriziomello) [inline expected/cnpj_1.out]: This file should be removed.

File: sql/cnpj.sql [hunk 6]
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE empresa(cnpj cnpj);
```
Comment (fabriziomello) [inline sql/cnpj.sql]: Missing index test. See the CPF test case.

File: sql/cnpj.sql [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+\d empresa
+INSERT INTO empresa VALUES (cnpj '00000000000000');
```
Comment (fabriziomello) [inline sql/cnpj.sql]: Missing implict cast test. See CPF test case.

File: sql/cnpj.sql [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+SELECT * FROM empresa;
```
Comment (fabriziomello) [inline sql/cnpj.sql]: Missing ordering test. See CPF test case.

File: expected/cnpj.out [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+ Column | Type | Collation | Nullable | Default 
+--------+------+-----------+----------+---------
+ cnpj   | cnpj |           |          | 
+
+INSERT INTO empresa VALUES (cnpj '00000000000000');
+SELECT * FROM empresa;
+        cnpj        
+--------------------
+ 00.000.000/0000-00
+(1 row)
```
Comment (fabriziomello) [inline expected/cnpj.out]: Is this a valid CNPJ?? I guess we should check for all digits equal and forbid it, no??

File: expected/cnpj_1.out (hunk 5 above)
Comment (azlev) [reply expected/cnpj_1.out]: This file is used by pg_regress in postgres 11, and cnpj.out to 12+. This is necessary right now due different error messages in postgres (ERROR:  invalid input syntax for integer: "33000.167/0661-29" vs. ERROR:  invalid input syntax for type bigint: "33000.167/0661-29") coming from postgres.

If there is a better way to do this, let me know.

File: expected/cnpj.out (hunk 9 above)
Comment (azlev) [reply expected/cnpj.out]: I misread the wikipedia page: https://pt.wikipedia.org/wiki/Cadastro_Nacional_da_Pessoa_Jur%C3%ADdica (Apesar de ser algo contra-intuitivo, existe um CNPJ com raiz 00.000.000 (ZERO), sendo que o mesmo é pertencente ao Banco do Brasil.), so it's not all zeroes, is just the number. I'll change the tests

About equal digits, I couldn't find any reference regarding this restriction.

File: expected/cnpj_1.out (hunk 5 above)
Comment (azlev) [reply expected/cnpj_1.out]: Ok, I found the documentation: https://github.com/postgres/postgres/blob/master/src/test/regress/pg_regress.c#L1327

the _1 is an alternative expectfile

**Review by fabriziomello — APPROVED** [review]

**Pull request merged**